presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java

```diff
@@ -0,0 +1,396 @@
+    {
+        tester.assertThat(new ReorderJoins(new CostComparator(1, 1, 1)))
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.values(new PlanNodeId("valuesA"), ImmutableList.of(p.symbol("A1", BIGINT)), TWO_ROWS),
+                                p.values(new PlanNodeId("valuesB"), ImmutableList.of(p.symbol("B1", BIGINT)), TWO_ROWS),
+                                ImmutableList.of(),
+                                ImmutableList.of(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT)),
+                                Optional.empty()))
+                .withStats("valuesA", PlanNodeStatsEstimate.builder()
+                        .setOutputRowCount(10000)
+                        .addSymbolStatistics(ImmutableMap.of(new Symbol("A1"), new SymbolStatsEstimate(0, 100, 0, 640000, 100)))
+                        .build())
+                .withStats("valuesB", PlanNodeStatsEstimate.builder()
+                        .setOutputRowCount(10000)
+                        .addSymbolStatistics(ImmutableMap.of(new Symbol("B1"), new SymbolStatsEstimate(0, 100, 0, 640000, 100)))
+                        .build())
+                .doesNotFire();
+    }
+
+    @Test
+    public void testDoesNotFireWithNoStats()
+    {
+        tester.assertThat(new ReorderJoins(new CostComparator(1, 1, 1)))
+                .withStatsCalculator(new UnknownStatsCalculator())
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.tableScan(ImmutableList.of(p.symbol("A1", BIGINT)), ImmutableMap.of(p.symbol("A1", BIGINT), new TestingColumnHandle("A1"))),
+                                p.values(new PlanNodeId("valuesB"), p.symbol("B1", BIGINT)),
+                                ImmutableList.of(new JoinNode.EquiJoinClause(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT))),
+                                ImmutableList.of(p.symbol("A1", BIGINT)),
+                                Optional.empty()))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testDoesNotFireForNonDeterministicFilter()
+    {
+        tester.assertThat(new ReorderJoins(new CostComparator(1, 1, 1)))
+                .on(p ->
+                        p.join(
+                                INNER,
+                                p.values(new PlanNodeId("valuesA"), p.symbol("A1", BIGINT)),
+                                p.values(new PlanNodeId("valuesB"), p.symbol("B1", BIGINT)),
+                                ImmutableList.of(new JoinNode.EquiJoinClause(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT))),
+                                ImmutableList.of(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT)),
+                                Optional.of(new ComparisonExpression(ComparisonExpressionType.LESS_THAN, p.symbol("A1", BIGINT).toSymbolReference(), new FunctionCall(QualifiedName.of("random"), ImmutableList.of())))))
```

Comment:
this example is not realistic because this filter can be pushed down under the join; it might be better to use non-deterministic filter that uses columns from both sides of the join

Reply:
I believe non-deterministic filters are never pushed down